KubeArmor/eventAuditor/ebpfManager.go

```diff
@@ -0,0 +1,366 @@
+		return fmt.Errorf("map %v not initialized", kaMap.Name)
+	}
+
+	if m.IsPinned() {
+		err1 = m.Unpin(m.PinPath())
+	}
+
+	if !bm.closeObjIfPossible(kaMap.FileName) {
+		// this is a straightforward approach as kubearmor/libbpf
+		// still does not support unloading specific maps
+		err2 = syscall.Close(m.FD())
+	}
+
+	delete(bm.maps, kaMap.Name)
+
+	return AppendErrors(err1, err2)
+}
+
+// InitProgram Function
+func (bm *KABPFManager) InitProgram(kaProg KABPFProg) error {
+	var o *lbpf.KABPFObject
+	var p *lbpf.KABPFProgram
+	var err error
+
+	if p = bm.getProg(kaProg.Name); p != nil {
+		return fmt.Errorf("program %v already initialized", kaProg.Name)
+	}
+
+	if o = bm.getObj(kaProg.FileName); o == nil {
+		if o, err = openAndLoadObj(bm.objsProgsPath, kaProg.FileName); err != nil {
+			return err
+		}
+	}
+
+	if p, err = o.FindProgramByName(string(kaProg.Name)); err != nil {
+		bm.closeObjIfPossible(kaProg.FileName)
+		return err
+	}
+
+	bm.progs[kaProg.Name] = p
+	bm.objs[kaProg.FileName] = o
+
+	return nil
+}
+
+// DestroyProgram Function
+func (bm *KABPFManager) DestroyProgram(kaProg KABPFProg) error {
+	var p *lbpf.KABPFProgram
+
+	if p = bm.getProg(kaProg.Name); p == nil {
+		return fmt.Errorf("program %v not initialized", kaProg.Name)
+	}
+
+	if !bm.closeObjIfPossible(kaProg.FileName) {
+		// this is a straightforward approach as kubearmor/libbpf
+		// still does not support unloading specific programs
+		if err := syscall.Close(p.FD()); err != nil {
+			return err
+		}
+	}
+
+	delete(bm.progs, kaProg.Name)
+
+	return nil
+}
+
+// AttachTracepoint Function
+func (bm *KABPFManager) AttachProgram(kaProg KABPFProg) error {
+	var p *lbpf.KABPFProgram
+	var l *lbpf.KABPFLink
+	var err error
+
+	if p = bm.getProg(kaProg.Name); p == nil {
+		return fmt.Errorf("program %v not initialized", kaProg.Name)
+	}
+
+	if l = bm.getLink(kaProg.EventName); l != nil {
+		return fmt.Errorf("program %v already attached to %v", kaProg.Name, kaProg.EventName)
+	}
+
+	switch kaProg.EventType {
+	case lbpf.KABPFLinkTypeTracepoint:
+		if l, err = p.AttachTracepoint(string(kaProg.EventName)); err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("undefined event type %v", kaProg.EventType)
+	}
+
+	bm.links[kaProg.EventName] = l
+
+	return nil
+}
+
+// DestroyProgram Function
+func (bm *KABPFManager) DetachProgram(kaProg KABPFProg) error {
+	var p *lbpf.KABPFProgram
+	var l *lbpf.KABPFLink
+	var err error
+
+	if p = bm.getProg(kaProg.Name); p == nil {
+		return fmt.Errorf("program %v not initialized", kaProg.Name)
+	}
+
+	if l = bm.getLink(kaProg.EventName); l == nil {
+		return fmt.Errorf("program %v not attached to %v", kaProg.Name, kaProg.EventName)
+	}
+
+	// https://github.com/kubearmor/libbpf/issues/18
+	if err = l.Detach(); err != nil {
+		return err
+	}
+
+	delete(bm.links, kaProg.EventName)
+
+	return err
+}
+
+// isObjCloseable Function
+func (bm *KABPFManager) isObjCloseable(o *lbpf.KABPFObject) bool {
```

Comment:
This function is a rough implementation. We can change it later to check an object reference counter.